lib/rgit/cli.rb

```diff
@@ -1,31 +1,234 @@
-            when Fixnum
-              result << key_str << "=#{v}"
-            when Hash
-              v.each { |key, val|
-                result << key_str << " #{key}=#{val}"
-              }
+    def self.extract_branches(path)
+      res = {}
+      if path.directory?
+        path.children.each do |child|
+          name = child.basename.to_s
+          if name.start_with?("%")
+            res[:sitting_branch] = child.to_s.split(".git/refs/heads/", 2)[1]
+          else
+            res[name] = extract_branches(child)
           end
+        end
+      end
+      res
+    end
+
+    def groups
+      @groups.keys
+    end
+
+    def [](group)
+      structure = @groups[group]
+      flat_hash(structure).values
+    end
+
+    def branch_for_path(path)
+      branch = nil
+      so_far = @groups[@current_group]
+      path.realpath.relative_path_from(@dir).descend do |f|
+        file_str = f.basename.to_s
+        sitting_branch = so_far[:sitting_branch]
+        branch = sitting_branch if sitting_branch
+        so_far = so_far[file_str]
+      end
+      branch
+    end
+
+    def path_for_branch(branch_fullname)
+      # branchname is a fully qualified path from top git repo
+      # + other info
+      # i.e. #{group}/full/path/to/#{sub_branch}
+      # unless it's a group-level branch in which case
+      # #{group}/%group
+      group, branch = branch_fullname.split("/", 2)
+      path = branch.rpartition("%")[0]
+      @dir + path
+    end
+  end
+
+  class Grafts
+    attr_accessor :location
+    def initialize(location)
+      @location = location
+      @grafts = {}
+    end
+
+    def [](key)
+      @grafts[key]
+    end
+
+    def []=(key, value)
+      @grafts[key] = value
+    end
+
+    def self.from_git(directory)
+      location = directory + ".git/info/grafts"
+      res = new(location)
+      location.open.readlines.each do |line|
+        # Shas are 40 characters, first sha is child
+        # TODO: HANDLE GRAFT FILES W/ MORE THAN ONE FAKE PARENT FOR A COMMIT
+        key, value = line.gsub(" ", "").scan(/.{40}/)
+        res[key] = value
+      end
+      res
+    end
+
+    def self.append_to_grafts!(directory, child_sha, parent_sha)
+      graft_dir = directory + ".git/info"
+      FileUtils.mkdir_p(graft_dir)
+
+      if (graft_dir + "grafts").exist?
+        grafts = from_git(directory)
+        grafts[child_sha] = parent_sha
+        grafts.write_to_disk!
+      else
+        File.open(graft_dir + "grafts", "w") { |file|
+          file.puts "#{child_sha} #{parent_sha}"
         }
-        result
       end
-    }
+    end
+
+    def self.graft_to_base!(directory, child_sha)
+      parent_sha = File.open(
+          directory + ".git/refs/heads/@rgit-base-for-graft",
+          &:readline).strip
+      append_to_grafts!(directory, child_sha, parent_sha)
+    end
+
+    def write_to_disk!
+      FileUtils.mkdir_p(@location.dirname)
+      File.open(@location, 'w') do |f|
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.